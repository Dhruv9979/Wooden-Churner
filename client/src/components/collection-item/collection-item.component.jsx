import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { addItem } from '../../redux/cart/cart.actions';

import {
    CollectionItemContainer,
    CollectionFooterContainer,
    AddButton,
    BackgroundImage,
    NameContainer,
    PriceContainer,
    TextContainer,
    TestOpacity,
} from './collection-item.styles';

export const CollectionItem = ({ item, addItem, match, history }) => {
    const { name, price, imageUrl, routeName, type, stock } = item;
    const outOfStock = 'Out Of Stock';
    const routingProducts = routeName
        ? match.url.indexOf(type) > -1
            ? `${match.url}/${routeName}`
            : `${match.url}/${type}/${routeName}`
        : `${match.url}`;
    var rupees ='\u20B9'; 

    const product = (
        <CollectionItemContainer>
            {stock === outOfStock ? (
                <BackgroundImage
                    to={routingProducts}
                    className="image"
                    imageUrl={imageUrl}
                >
                    <TestOpacity></TestOpacity>
                    <TextContainer>{stock}</TextContainer>
                </BackgroundImage>
            ) : (
                <BackgroundImage
                    to={routingProducts}
                    className="image"
                    imageUrl={imageUrl}
                />
            )}
            <CollectionFooterContainer>
                <NameContainer>{name}</NameContainer>
                { typeof (price) === 'number' &&<PriceContainer>{rupees} {price}</PriceContainer>}
            </CollectionFooterContainer>
            {routeName ? (
                ''
            ) : stock === outOfStock ? (
                ''
            ) : (
                <AddButton onClick={() => addItem(item)}>ADD TO CART</AddButton>
            )}
        </CollectionItemContainer>
    );

    return product;
};

const mapDispatchToProps = (dispatch) => ({
    addItem: (item) => dispatch(addItem(item)),
});

export default withRouter(connect(null, mapDispatchToProps)(CollectionItem));
