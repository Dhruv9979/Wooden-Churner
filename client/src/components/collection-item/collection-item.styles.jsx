import styled from 'styled-components';
import { Link } from 'react-router-dom';
import CustomButton from '../custom-button/custom-button.component';

export const CollectionItemContainer = styled.div`
    width: 22vw;
    display: flex;
    flex-direction: column;
    height: 350px;
    align-items: center;
    position: relative;
    &:hover {
        .image {
            opacity: 0.8;
        }
        button {
            opacity: 0.85;
            display: flex;
        }
    }

    @media screen and (max-width: 800px) {
        width: 40vw;
        &:hover {
            .image {
                opacity: unset;
            }
            button {
                opacity: unset;
            }
        }
    }
`;

export const AddButton = styled(CustomButton)`
    width: 80%;
    opacity: 0.7;
    position: absolute;
    top: 255px;
    display: none;

    @media screen and (max-width: 800px) {
        top: 210px;
        display: block;
        opacity: 0.9;
        min-width: unset;
        padding: 0 10px;
        font-size: 14px;
        font-weight: bolder;
    }
`;

AddButton.displayName = 'AddButton';

export const BackgroundImage = styled(Link)`
    width: 100%;
    height: 90%;
    background-repeat: no-repeat;
    background-size: contain;
    background-position: center;
    margin-bottom: 5px;
    position: relative;
    background-image: ${({ imageUrl }) => `url(${imageUrl})`};
`;

BackgroundImage.displayName = 'BackgroundImage';

export const TextContainer = styled.div`
    position: relative;
    z-index: 1;
    font-size: 35px;
    text-transform: uppercase;
    top: 50%;
    left: 5%;
    transform: rotate(65deg);
`;

export const TestOpacity = styled.div`
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    background-color: #fff;
    opacity: 0.8;
`;

TestOpacity.displayName = 'TestOpacity';

export const CollectionFooterContainer = styled.div`
    width: 100%;
    height: 10%;
    display: flex;
    justify-content: space-between;
    padding: 0 45px;
    font-size: 15px;
    color: #762522;
    @media screen and (max-width: 800px) {
        padding: 0 15px;
        height: 25%;
    }
`;

CollectionFooterContainer.displayName = 'CollectionFooterContainer';

export const NameContainer = styled.span`
    width: 100%;
    margin-bottom: 15px;
    text-align: center;

    @media screen and (max-width: 800px) {
        width: 70%;
        text-align: left;
    }
`;

NameContainer.displayName = 'NameContainer';

export const PriceContainer = styled.span`
    width: 30%;
    text-align: right;
`;

PriceContainer.displayName = 'PriceContainer';
