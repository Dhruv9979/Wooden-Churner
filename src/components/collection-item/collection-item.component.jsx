import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { Link } from "react-router-dom";

import CustomButton from "../custom-button/custom-button.component";
import { addItem } from "../../redux/cart/cart.actions";

import "./collection-item.styles.scss";

//Check for routeName.
// If yes, check for type. If no then match.url.
//If yes for type then add type to route. If no, then match.url/routeName
// routeName ? (type ? `${match.url}/${type}/${routeName}` : `${match.url}/${routeName}`) : `${match.url}`


const CollectionItem = ({ item, addItem, match, history }) => {
	const { name, price, imageUrl, routeName, type } = item;
	
	const routingProducts = routeName
		? match.url.indexOf(type) > -1
			? `${match.url}/${routeName}`
			: `${match.url}/${type}/${routeName}`
		: `${match.url}`;

	return (
		<div className="collection-item">
			<Link
				to={routingProducts}
				className="image"
				style={{
					backgroundImage: `url(${imageUrl})`,
				}}
			/>
			<div className="collection-footer">
				<span className="name">{name}</span>
				<span className="price">&#x20B9;{price}</span>
			</div>
			{routeName ? (
				""
			) : (
				<CustomButton inverted onClick={() => addItem(item)}>
					ADD TO CART
				</CustomButton>
			)}
		</div>
	);
};

const mapDispatchToProps = (dispatch) => ({
	addItem: (item) => dispatch(addItem(item)),
});

export default withRouter(connect(null, mapDispatchToProps)(CollectionItem));
