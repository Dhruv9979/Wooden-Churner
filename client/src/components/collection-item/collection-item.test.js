import React from "react";
import { shallow, mount } from "enzyme";
import { CollectionItem } from "./collection-item.component";

describe("CollectionItem component", () => {
    let wrapper;
    let mockAddItem;
    let mockMatch;
    let imageUrl = "https://i.ibb.co/bRYN3Bx/Almond-Oil.png";
    let mockName = "Almond Oil";
    let mockPrice = "10 - 20";
    let mockRouteName = "oils";

    beforeEach(() => {
        mockMatch = {
            url: "/shop/oils",
        };
        const mockProps = {
            item: {
                imageUrl: imageUrl,
                price: mockPrice,
                name: mockName,
                routeName: mockRouteName,
            },
            match: mockMatch,
        };

        wrapper = shallow(<CollectionItem {...mockProps} />);
    });

    it("should render CollectionItem Component", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should not render AddButton on the Collection Page", () => {
        expect(wrapper.find("AddButton").exists()).toBeFalsy();
    });

    it('should render AddButton on Product Page', () => {
        mockAddItem = jest.fn();
        const mockProps = {
            item: {
                imageUrl: imageUrl,
                price: mockPrice,
                name: mockName,
                routeName: undefined,
            },
            addItem: mockAddItem,
            match: mockMatch,
        };
    
        wrapper = shallow(<CollectionItem {...mockProps} />);
    
        wrapper.find("AddButton").simulate("click");
    
        expect(mockAddItem).toHaveBeenCalled();
    });

    it("should render imageUrl as a prop on BackgroundImage", () => {
        expect(wrapper.find("BackgroundImage").prop("imageUrl")).toBe(imageUrl);
    });

    it("should render name prop in NameContainer", () => {
        expect(wrapper.find("NameContainer").text()).toBe(mockName);
    });

    it("should render price prop in PriceContainer", () => {
        const price = 'â‚¹' + mockPrice;
        expect(wrapper.find("PriceContainer").text()).toBe(price);
    });
});
