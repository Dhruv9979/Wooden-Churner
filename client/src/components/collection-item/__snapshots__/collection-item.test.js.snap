// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionItem component should render CollectionItem Component 1`] = `ShallowWrapper {}`;
